libvast/vast/flatbuffer.hpp

```diff
@@ -0,0 +1,350 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2021 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#pragma once
+
+#include "vast/fwd.hpp"
+
+#include "vast/chunk.hpp"
+
+#include <flatbuffers/flatbuffers.h>
+#include <fmt/core.h>
+
+#include <compare>
+
+namespace vast {
+
+/// A wrapper class around a FlatBuffers table that allows for sharing the
+/// lifetime with the chunk containing the table.
```

Comment:
Please mention that it is reference-counted and copyable.

I also wonder if it would make sense to make a separate `root_flatbuffer` class instead of the boolean template argument, but that probably goes too far for this PR.

Reply:
I thought about having the `root_flatbuffer` as well, but it would really add a lot of code and as a developer I want the root variant to be the default, not the other way around. I think given that we almost always use `auto` we have an easy time using this in our code. We could probably add this alias for ease of use:

```cpp
template <class Table>
using child_flatbuffer = flatbuffer<Table, false>;
```

But I don't see the immediate win in adding that now. Future uses of this will show what we really need, and we can continually improve this I think.